docs/internal.rst

```diff
@@ -6,6 +6,68 @@ Intermediate representation
 Use ``ti.init(print_ir=True)`` to print IR on the console.
 
 
+Data Structure Node Organization
+--------------------------------
+
+The internal organization of data structure nodes can be confusing.
+It is important to distinguish the concept of **containers**, **cells** and **components**.
+
+- A **container** can have multiple **cells**. The numbers of **cells** are recommended to be powers of two.
+- A **cell** can have multiple **components**.
+- Each **component** is a **container** of a lower-level SNode.
+
+Consider the following example:
+
+.. code-block:: python
+
+    # misc/listgen_demo.py
+
+    x = ti.var(ti.i32)
+    y = ti.var(ti.i32)
+    z = ti.var(ti.i32)
+
+    S0 = ti.root
+    S1 = S0.pointer(ti.i, 4)
+
+    S2 = S1.dense(ti.i, 8)
+    S2.place(x, y) # S3: y; S4: z
+
+    S5 = S1.dense(ti.i, 8)
+    S5.place(z) # S6: z
+
+
+- The whole data structure is a ``S0root`` **container**, containing
+
+  - ``1x`` ``S0root`` **cell**, each has only one **component**, which is
```

Comment:
For the type name of cells, do you think it's more obvious if we append `_ch`, e.g. `S0root_ch`? I think this is the actual struct name on the backends?

Reply:
I actually feel a little reluctant here to use `_ch` because it reminds people of **children**, which is ambiguous since it can either be a `cell` of a `container`, or an SNode children of another SNode...

Once we reach a consensus on the names, I'll refactor the struct compilers.